baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,361 @@
+            headerString += "|"
+        headerWidth = len(headerString)
+
+        headerBar = "+" + "-" * (headerWidth - 2) + "+"
+        print(headerBar)
+        print(headerString)
+        print(headerBar)
+
+    def printData(self, iters: Optional[Union[int, Iterable[int]]] = None) -> None:
+        """Print a selection of lines from the history
+
+        Parameters
+        ----------
+        iters : int or Iterable of ints, optional
+            Iteration numbers to print, by default only the last iteration will be printed
+        """
+        if iters is None:
+            iters = [-1]
+        elif isinstance(iters, int):
+            iters = [iters]
+
+        if max(iters) >= self.iter:
+            raise ValueError(
+                f"Requested iteration {max(iters)} is not in the history, only {self.iter} iterations in history"
+            )
+        if min(iters) < -self.iter:
+            raise ValueError(
+                f"Requested iteration {min(iters)} is not in the history, only {self.iter} iterations in history"
+            )
+
+        for i in iters:
+            lineString = "|"
+            for var in self.variablesToPrint:
+                if self.data[var][i] is None:
+                    lineString += self.variables[var]["headerFormat"].format("-")
+                else:
+                    valueString = self.variables[var]["format"].format(self.data[var][i])
+                    lineString += self.variables[var]["headerFormat"].format(valueString)
+                lineString += "|"
+            print(lineString)
+
+    def save(self, fileName: str) -> None:
+        """Write the solution history to a pickle file
+
+        Only the data dictionary is saved
+
+        Parameters
+        ----------
+        fileName : str
+            File path to save the solution history to, file extension not required, will be ignored if supplied
+        """
+        base = os.path.splitext(fileName)[0]
+        fileName = base + ".pkl"
+
+        dataToSave = {"data": self.getData(), "metadata": self.getMetadata()}
+        with open(fileName, "wb") as file:
+            pickle.dump(dataToSave, file, protocol=pickle.HIGHEST_PROTOCOL)
+
+    def getData(self) -> Dict[str, List]:
+        """Get the recorded data
+
+        Returns
+        -------
+        dict
+            Dictionary of recorded data
+        """
+        return copy.deepcopy(self.data)
+
+    def getMetadata(self) -> Dict[str, Any]:
+        """Get the recorded metadata
+
+        Returns
+        -------
+        dict
+            Dictionary of recorded metadata
+        """
+        return copy.deepcopy(self.metadata)
+
+    def getVariables(self) -> List[str]:
+        """Get the recorded variables
+
+        Returns
+        -------
+        dict
+            Dictionary of recorded variables
+        """
+        return list(self.variables.keys())
+
+    @property
+    def variablesToPrint(self) -> List[str]:
```

Comment:
why are you using this approach? Is that to make sure that `variablesToPrint` is always up to date?

Reply:
Yeah, it would be hard to know at which point to compute `variablesToPrint` otherwise (not really any way for the solver history to know when the user has stopped adding variables to be printed